src/Language/Syntax.hs

```diff
@@ -86,14 +88,22 @@ instance Show Btype where
   show Player = "Player"
   show Position = "Position"
   show Positions = "Positions"
+  show Undef = "?"
 
 -- | Xtypes are sum types, but restricted by the semantics to only contain Symbols after the atomic type.
-data Xtype = X Btype [Btype]
-   deriving (Eq) 
+data Xtype = X Btype (S.Set Name)
+
+-- This is a potential source of very confusing bugs. beware. Also this instance is not symmetric, which is bad.
+instance Eq Xtype where
```

Comment:
I'm confused about this Eq instance, but you know where and how its used and I'll trust that. 

Why would we want `X (Symbol "T2") (S.fromList ["T"]) == X Player (S.fromList ["T2"])` to be `True` though?

Reply:
the initial reason was because those two types can be resolved into a single type (Player | T | T2). For example, with `f : Int -> T2 | T` and `g : Int -> Player | T2`, my understanding is that this function should be accepted by the typechecker:
```
g : Int -> Player | T1 | T2
g(x) = if x == 0 then f(x) else g(x)
```

Using the `Eq` instance for this was probably a bad idea on my part. I will refactor this into `Typechecker.hs` and probably just leave the derived part in. These are two questions for the next meeting though:
1. Is the function `h` I defined above type correct?
2. Should promotion from `Int` to `Int | T` be allowed by the typechecker?